combinetf2/fitter.py

```diff
@@ -539,7 +537,132 @@ def global_impacts_parms(self):
+            pd2ldbeta2 = tf.linalg.LinearOperatorFullMatrix(
+                pd2ldbeta2_matrix, is_self_adjoint=True
+            )
+        else:
+            # pd2ldbeta2 is diagonal, so we can use gradient instead of jacobian
+            pd2ldbeta2_diag = t2.gradient(pdldbeta, self.ubeta)
+            pd2ldbeta2 = tf.linalg.LinearOperatorDiag(
+                pd2ldbeta2_diag, is_self_adjoint=True
+            )
+        return pd2ldbeta2
+
+    def _dxdvars(self):
+        with tf.GradientTape() as t2:
+            t2.watch([self.theta0, self.nobs, self.beta0])
+            with tf.GradientTape() as t1:
+                t1.watch([self.theta0, self.nobs, self.beta0])
+                val = self._compute_loss()
+            grad = t1.gradient(val, self.x)
+        pd2ldxdtheta0, pd2ldxdnobs, pd2ldxdbeta0 = t2.jacobian(
+            grad, [self.theta0, self.nobs, self.beta0], unconnected_gradients="zero"
+        )
+
+        # cov is inverse hesse, thus cov ~ d2xd2l
+        dxdtheta0 = -self.cov @ pd2ldxdtheta0
+        dxdnobs = -self.cov @ pd2ldxdnobs
+        dxdbeta0 = -self.cov @ pd2ldxdbeta0
+
+        return dxdtheta0, dxdnobs, dxdbeta0
+
+    def _chi2(
+        self,
+        fun,
+        ndf_reduction=0,
+        profile=False,
+    ):
+
+        with tf.GradientTape() as t:
+            t.watch([self.theta0, self.nobs, self.beta0])
+            expected = self._compute_expected(
+                fun,
+                inclusive=True,
+                profile=profile,
+                full=False,
+                need_observables=True,
+            )
+            observed = fun(None, self.nobs)
+            residuals = expected - observed
+
+            residuals_flat = tf.reshape(residuals, (-1,))
+        pdresdx, pdresdtheta0, pdresdnobs, pdresdbeta0 = t.jacobian(
+            residuals_flat,
+            [self.x, self.theta0, self.nobs, self.beta0],
+            unconnected_gradients="zero",
+        )
+
+        if profile:
+            # apply chain rule to take into account correlations with the fit parameters
+            dxdtheta0, dxdnobs, dxdbeta0 = self._dxdvars()
+
+            dresdtheta0 = pdresdtheta0 + pdresdx @ dxdtheta0
+            dresdnobs = pdresdnobs + pdresdx @ dxdnobs
+            dresdbeta0 = pdresdbeta0 + pdresdx @ dxdbeta0
+
+            var_theta0 = tf.where(
+                self.indata.constraintweights == 0.0,
+                tf.zeros_like(self.indata.constraintweights),
+                tf.math.reciprocal(self.indata.constraintweights),
+            )
+
+            res_cov = dresdtheta0 @ (var_theta0[:, None] * tf.transpose(dresdtheta0))
+        else:
+            # full dependency in partial derivatives
+            dresdnobs = pdresdnobs
+            dresdbeta0 = pdresdbeta0
+
+            res_cov = pdresdx @ tf.matmul(self.cov, pdresdx, transpose_b=True)
+
+        res_cov_stat = dresdnobs @ (self.nobs[:, None] * tf.transpose(dresdnobs))
```

Comment:
This is not valid in the case of externalCovariance
(would have to use the data covariance here)

The logic can probably be simplified by using LinearOperatorDiag for the diagonal case to keep things more similar with vs without the externalCovarince

Reply:
Fixed, I think here without the LinearOperatorDiag is simpler here since we have to work with the inverse of the data covariance in one case and the data variance in the other